process/transactions/tokens.go

```diff
@@ -0,0 +1,240 @@
+	issueSemiFungibleESDTFunc = "issueSemiFungible"
+	issueNonFungibleESDTFunc  = "issueNonFungible"
+)
+
+type tokensProcessor struct {
+	issueMethods           map[string]struct{}
+	argumentParserExtended *argumentsParserExtended
+	selfShardID            uint32
+}
+
+func newTokensProcessor(selfShardID uint32) *tokensProcessor {
+	argsParser := parsers.NewCallArgsParser()
+
+	return &tokensProcessor{
+		selfShardID: selfShardID,
+		issueMethods: map[string]struct{}{
+			issueFungibleESDTFunc:     {},
+			issueSemiFungibleESDTFunc: {},
+			issueNonFungibleESDTFunc:  {},
+		},
+		argumentParserExtended: newArgumentsParser(argsParser),
+	}
+}
+
+func (tp *tokensProcessor) searchForTokenIssueTransactions(txs []*data.Transaction, timestamp uint64) []*data.TokenInfo {
+	if tp.selfShardID != core.MetachainShardId {
+		return []*data.TokenInfo{}
+	}
+
+	tokensData := make([]*data.TokenInfo, 0)
+	for _, tx := range txs {
+		funcName, okSCR, ok := tp.isIssueTxSuccess(tx.Data, tx.Sender, tx.Nonce, tx.SmartContractResults)
+		if !ok {
+			continue
+		}
+
+		tokenInfo, ok := tp.extractTokenInfo(tx.Data, okSCR)
+		if !ok {
+			continue
+		}
+
+		if funcName == issueFungibleESDTFunc {
+			tokenInfo.Token = tp.searchTokenIdentifierFungibleESDT(tx.Sender, tx.SmartContractResults)
+		}
+
+		tokenInfo.Timestamp = time.Duration(timestamp)
+		tokensData = append(tokensData, tokenInfo)
+	}
+
+	return tokensData
+}
+
+func (tp *tokensProcessor) searchForTokenIssueScrs(scrs []*data.ScResult, timestamp uint64) []*data.TokenInfo {
+	if tp.selfShardID != core.MetachainShardId {
+		return []*data.TokenInfo{}
+	}
+
+	tokensData := make([]*data.TokenInfo, 0)
+	for _, scr := range scrs {
+		funcName, resScr, ok := tp.isIssueSCRSuccess(scr.Data, scr.Sender, scrs)
+		if !ok {
+			continue
+		}
+
+		tokenInfo, ok := tp.extractTokenInfo(scr.Data, resScr)
+		if !ok {
+			continue
+		}
+
+		tokenInfo.Timestamp = time.Duration(timestamp)
+
+		if funcName != issueFungibleESDTFunc {
+			tokensData = append(tokensData, tokenInfo)
+			continue
+		}
+
+		identifier := tp.searchTokenIdentifierFungibleESDT(scr.Sender, scrs)
+		if identifier == "" {
+			continue
+		}
+		tokenInfo.Issuer = scr.Sender
+		tokenInfo.Token = identifier
+
+		tokensData = append(tokensData, tokenInfo)
+	}
+
+	return tokensData
+}
+
+func (tp *tokensProcessor) searchTokenIdentifierFungibleESDT(sender string, scrs []*data.ScResult) string {
+	for _, scr := range scrs {
+		isReceiverAndSenderOk := scr.Receiver == sender && scr.Sender == metachainESDTAddress
+		if !isReceiverAndSenderOk {
+			continue
+		}
+		if scr.Nonce != 0 {
+			continue
+		}
+
+		funcName, args, err := tp.argumentParserExtended.ParseData(string(scr.Data))
+		if err != nil {
+			continue
+		}
+
+		if funcName != core.BuiltInFunctionESDTTransfer || len(args) < 2 {
+			continue
+		}
+
+		return string(args[0])
+	}
+
+	return ""
+}
+
+func (tp *tokensProcessor) extractTokenInfo(dataField []byte, scr *data.ScResult) (*data.TokenInfo, bool) {
```

Comment:
could have returned error instead of bool

Reply:
in this situation, I prefer to return the bool value because in not all cases we can extract token information . For example, if we have a transaction that failed (almost all check have passed but we don't have a SCR with @ok, in this case, I return 'false' , have no error here)